### PR TITLE
Json schema based mock verifier

### PR DESCRIPTION
**Pull request by jaceko**

### CODE REVIEW

File: test-utils/test-utils-core/src/main/java/uk/gov/justice/services/test/utils/core/mock/JsonSchemaValidatingMockMaker.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        return delegate.createMock(settings, handler);
+    }
+
+    @Override
+    public MockHandler getHandler(final Object object) {
+        return delegate.getHandler(object);
+    }
+
+    @Override
+    public void resetMock(final Object object, final MockHandler mockHandler, final MockCreationSettings mockCreationSettings) {
```
Comment (mapingo) [inline test-utils/test-utils-core/src/main/java/uk/gov/justice/services/test/utils/core/mock/JsonSchemaValidatingMockMaker.java]: This method does not seem to be tested.

File: core/src/test/java/uk/gov/justice/services/core/json/JsonSchemaLoaderTest.java [hunk 2]
```diff
@@ -34,8 +30,14 @@ public void shouldReturnSchemaFromClasspath() {
...
     @Test
     public void shouldLogSchemaName() throws Exception {
+        final TestLogAppender testLogAppender = TestLogAppender.activate();
+
         loader.loadSchema("test-schema");
-        verify(logger).trace("Loading schema {}", "/json/schema/test-schema.json");
+        testLogAppender.deactivate();
+        final LoggingEvent logEntry = testLogAppender.firstLogEntry();
+        Assert.assertThat(logEntry.getLevel(), is(TRACE));
+        Assert.assertThat(logEntry.getMessage(), is("Loading schema /json/schema/test-schema.json"));
```
Comment (mapingo) [inline core/src/test/java/uk/gov/justice/services/core/json/JsonSchemaLoaderTest.java]: Assert can be static import

File: test-utils/test-utils-core/src/main/java/uk/gov/justice/services/test/utils/core/mock/JsonSchemaValidatingMockMaker.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+import org.mockito.mock.MockCreationSettings;
+import org.mockito.plugins.MockMaker;
+
+/**
+ * Mockito extension to test if payloads passed to {@link Sender} and returned by {@link Requester},
+ * conform to respective json schemas.
+ */
+public class JsonSchemaValidatingMockMaker implements MockMaker {
+
+    private final MockMaker delegate = new CglibMockMaker();
```
Comment (mapingo) [inline test-utils/test-utils-core/src/main/java/uk/gov/justice/services/test/utils/core/mock/JsonSchemaValidatingMockMaker.java]: delegate as variable name, maybe mockMaker would be better

File: test-utils/test-utils-core/src/main/java/uk/gov/justice/services/test/utils/core/mock/JsonSchemaValidatingMockMaker.java (hunk 1 above)
Comment (jaceko) [inline test-utils/test-utils-core/src/main/java/uk/gov/justice/services/test/utils/core/mock/JsonSchemaValidatingMockMaker.java]: It's tested indirectly by JsonSchemaValidatingMockTest

File: test-utils/test-utils-core/src/main/java/uk/gov/justice/services/test/utils/core/mock/JsonSchemaValidatingMockMaker.java (hunk 3 above)
Comment (jaceko) [inline test-utils/test-utils-core/src/main/java/uk/gov/justice/services/test/utils/core/mock/JsonSchemaValidatingMockMaker.java]: done

File: test-utils/test-utils-core/src/main/java/uk/gov/justice/services/test/utils/core/mock/JsonSchemaValidatingMockMaker.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+    @Override
+    public void resetMock(final Object object, final MockHandler mockHandler, final MockCreationSettings mockCreationSettings) {
+        delegate.resetMock(object, mockHandler, mockCreationSettings);
+    }
+
+    private void validateAgainstJsonSchema(final MethodInvocationReport methodInvocationReport) {
+
+        final Optional<JsonEnvelope> envelopeToBeValidated = envelopeToValidateFrom(methodInvocationReport);
+        if (envelopeToBeValidated.isPresent()) {
```
Comment (jaceko) [inline test-utils/test-utils-core/src/main/java/uk/gov/justice/services/test/utils/core/mock/JsonSchemaValidatingMockMaker.java]: done

File: core/src/test/java/uk/gov/justice/services/core/json/JsonSchemaLoaderTest.java (hunk 2 above)
Comment (jaceko) [inline core/src/test/java/uk/gov/justice/services/core/json/JsonSchemaLoaderTest.java]: done

File: test-utils/test-utils-core/src/test/java/uk/gov/justice/services/test/utils/core/mock/JsonSchemaValidatingMockTest.java [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+
+    public static class SendingHandler {
+        private Sender sender;
+
+
+        public SendingHandler(final Sender sender) {
+            this.sender = sender;
+        }
+
+        public void handle(JsonEnvelope envelope) {
```
Comment (matt-rich) [inline test-utils/test-utils-core/src/test/java/uk/gov/justice/services/test/utils/core/mock/JsonSchemaValidatingMockTest.java]: missing final, inconsistency, and the inner classes below

File: test-utils/test-utils-core/src/main/java/uk/gov/justice/services/test/utils/core/mock/JsonSchemaValidatingMockMaker.java [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    private void validateAgainstJsonSchema(final MethodInvocationReport methodInvocationReport) {
+        envelopeToValidateFrom(methodInvocationReport).ifPresent(this::validateAgainstJsonSchema);
+    }
+
+    private Optional<JsonEnvelope> envelopeToValidateFrom(final MethodInvocationReport methodInvocationReport) {
+
+        final Invocation invocation = (Invocation) methodInvocationReport.getInvocation();
+        if (invocationClassIs(invocation, Sender.class)) {
+            return envelopeOf(invocation.getArguments()[0]);
+        } else if (invocationClassIs(invocation, Requester.class) && methodInvocationReport.getReturnedValue() != null) {
```
Comment (matt-rich) [inline test-utils/test-utils-core/src/main/java/uk/gov/justice/services/test/utils/core/mock/JsonSchemaValidatingMockMaker.java]: Could we wrap this condition in a private method to make it clearer what it is.

File: test-utils/test-utils-core/src/main/java/uk/gov/justice/services/test/utils/core/mock/JsonSchemaValidatingMockMaker.java [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+            return envelopeOf(methodInvocationReport.getReturnedValue());
+        } else {
```
Comment (matt-rich) [inline test-utils/test-utils-core/src/main/java/uk/gov/justice/services/test/utils/core/mock/JsonSchemaValidatingMockMaker.java]: We can remove the else around the Optional.empty() and just drop into a return  Optional.empty() as the if and else both return to. <--- Spotted by Allan

File: test-utils/test-utils-core/src/test/java/uk/gov/justice/services/test/utils/core/mock/JsonSchemaValidatingMockTest.java (hunk 5 above)
Comment (jaceko) [inline test-utils/test-utils-core/src/test/java/uk/gov/justice/services/test/utils/core/mock/JsonSchemaValidatingMockTest.java]: fixed

File: test-utils/test-utils-core/src/main/java/uk/gov/justice/services/test/utils/core/mock/JsonSchemaValidatingMockMaker.java (hunk 6 above)
Comment (jaceko) [inline test-utils/test-utils-core/src/main/java/uk/gov/justice/services/test/utils/core/mock/JsonSchemaValidatingMockMaker.java]: fixed

File: test-utils/test-utils-core/src/main/java/uk/gov/justice/services/test/utils/core/mock/JsonSchemaValidatingMockMaker.java (hunk 7 above)
Comment (jaceko) [inline test-utils/test-utils-core/src/main/java/uk/gov/justice/services/test/utils/core/mock/JsonSchemaValidatingMockMaker.java]: fixed

**Pull request merged**